src/event.rs

```diff
@@ -0,0 +1,616 @@
+				);
+				let (payment_preimage, payment_secret) = match purpose {
+					PaymentPurpose::InvoicePayment { payment_preimage, payment_secret, .. } => {
+						(payment_preimage, Some(payment_secret))
+					}
+					PaymentPurpose::SpontaneousPayment(preimage) => (Some(preimage), None),
+				};
+				let mut payments = self.inbound_payments.lock().unwrap();
+				match payments.entry(payment_hash) {
+					hash_map::Entry::Occupied(mut e) => {
+						let payment = e.get_mut();
+						payment.status = PaymentStatus::Succeeded;
+						payment.preimage = payment_preimage;
+						payment.secret = payment_secret;
+					}
+					hash_map::Entry::Vacant(e) => {
+						e.insert(PaymentInfo {
+							preimage: payment_preimage,
+							secret: payment_secret,
+							status: PaymentStatus::Succeeded,
+							amount_msat: Some(amount_msat),
+						});
+					}
+				}
+				self.event_queue
+					.add_event(Event::PaymentReceived { payment_hash, amount_msat })
+					.expect("Failed to push to event queue");
+			}
+			LdkEvent::PaymentSent { payment_preimage, payment_hash, fee_paid_msat, .. } => {
+				let mut payments = self.outbound_payments.lock().unwrap();
+				for (hash, payment) in payments.iter_mut() {
+					if *hash == payment_hash {
+						payment.preimage = Some(payment_preimage);
+						payment.status = PaymentStatus::Succeeded;
+						log_info!(
+							self.logger,
+							"Successfully sent payment of {} millisatoshis{} from \
+								 payment hash {:?} with preimage {:?}",
+							payment.amount_msat.unwrap(),
+							if let Some(fee) = fee_paid_msat {
+								format!(" (fee {} msat)", fee)
+							} else {
+								"".to_string()
+							},
+							hex_utils::to_string(&payment_hash.0),
+							hex_utils::to_string(&payment_preimage.0)
+						);
+					}
+				}
+				self.event_queue
+					.add_event(Event::PaymentSuccessful { payment_hash })
+					.expect("Failed to push to event queue");
+			}
+			LdkEvent::PaymentFailed { payment_hash, .. } => {
+				log_info!(
+					self.logger,
+					"Failed to send payment to payment hash {:?}: exhausted payment retry attempts",
+					hex_utils::to_string(&payment_hash.0)
+				);
+
+				let mut payments = self.outbound_payments.lock().unwrap();
+				if payments.contains_key(&payment_hash) {
+					let payment = payments.get_mut(&payment_hash).unwrap();
+					assert_eq!(payment.status, PaymentStatus::Pending);
+					payment.status = PaymentStatus::Failed;
+				}
+				self.event_queue
+					.add_event(Event::PaymentFailed { payment_hash })
+					.expect("Failed to push to event queue");
+			}
+
+			LdkEvent::PaymentPathSuccessful { .. } => {}
+			LdkEvent::PaymentPathFailed { .. } => {}
+			LdkEvent::ProbeSuccessful { .. } => {}
+			LdkEvent::ProbeFailed { .. } => {}
+			LdkEvent::HTLCHandlingFailed { .. } => {}
+			LdkEvent::PendingHTLCsForwardable { time_forwardable } => {
+				let locked_runtime = self.tokio_runtime.read().unwrap();
+				if locked_runtime.as_ref().is_none() {
+					return;
+				}
```

Comment:
Should we panic instead?

Reply:
Mh, not sure, since I'd like to avoid a spurious panic when the user calls `LdkLite::stop()`. I think the current shutdown sequence *should* avoid this, but I'm not 100% sure. But maybe we should call `process_pending_htlc_forwards` on `LdkLite::start()`, similarly to the behavior in `ChannelManager::read()` (https://github.com/lightningdevkit/rust-lightning/blob/3d2c8f5937e884ff9769570c6de2acbfa15dcb4f/lightning/src/ln/channelmanager.rs#L7549-L7558)?

Reply:
IIUC, we should get this event again on restart when the `ChannelManager` is read?

I guess I'm thinking more if we may ever accidentally get into this state and fail to forward HTCLs. Who is expected to call `set_runtime`?

Reply:
The `set_runtime`/`drop_runtime` methods will be called upon calling `LdkLite::start`/`LdkLite::stop`, i.e., if we really have no runtime available anymore, we shouldn't continue operation, but leave the object in a state where it can be re-started.

I agree that the event would be re-generated upon *`read`*, however, note that the ChannelManager currently won't be re-read upon `start`, but only when the `LdkLite` object is initialized. So just to make sure we'd immediately start again where we left off, calling `process_pending_htlc_forwards` probably won't hurt?

Reply:
Yeah, likely. Tagging @TheBlueMatt to confirm.

Reply:
Well, at least the current shutdown sequence is a bit different and the main difference is that we *won't* drop and reinit all LDK objects on `stop()`/`start()`. I'd however suggest discussing this further when we come to #11, where we'll have the context.

Reply:
I agree it should be unreachable, but I'm still not convinced we want our code to ever panic in `release` build, as I fear that in case there were a bug the user would have an even harder time recovering if the app just spuriously fails rather than at least try to recover and resume operation. My reservations against any panics are also reinforced as downtime of a Lightning node is not only a matter of safety, but also of security.

As this is however more of a meta question, I now reverted to panicking here and in the wallet code with https://github.com/lightningdevkit/ldk-node/pull/10/commits/b2f3872d7f80e2de444c859ba10679ca5c066cb7.